Replaced shutil and tempfile packages with joblib in caching transformers

kirilldolmatov opened the pull request:
Replaced shutil and tempfile packages with joblib functions according to Memory usage in [official documentation](https://joblib.readthedocs.io/en/latest/auto_examples/memory_basic_usage.html)

#### What does this implement/fix? Explain your changes.
This PR simplifies using of caching transformers within a Pipeline. 

#### Any other comments?
This is my first PR.

kirilldolmatov committed:
replaced shutil and tempfile packages with joblib

Replaced shutil and tempfile packages with joblib functions according to Memory usage in [official documentation](https://joblib.readthedocs.io/en/latest/auto_examples/memory_basic_usage.html)

rth commented on examples/compose/plot_compare_reduction.py:
```diff
@@ -102,13 +102,11 @@
 # Create a temporary folder to store the transformers of the pipeline
-cachedir = mkdtemp()
-memory = Memory(location=cachedir, verbose=10)
+location = './cachedir'
```
Just `"cachedir"` for an OS independent path, otherwise LGTM. Thanks!

kirilldolmatov replied:
Done

rth reviewed (commented):
Looks good except for one comment below. Thanks @kirilldolmatov!

rth commented on examples/compose/plot_compare_reduction.py:
```diff
@@ -119,7 +117,7 @@
 
 # Delete the temporary cache before exiting
-rmtree(cachedir)
+memory.clear(warn=False)
```
Have you checked that this also removes the location folder?

kirilldolmatov replied:
yes, I've checked after calling 'clear()' location folder is removed. However, not main directory - 'cachedir' (in this case), but subdirectory - 'sklearn'.

rth replied:
OK, then we still want to remove the `cachedir` though.

kirilldolmatov replied:
I found smth that can meet our requirements: [clear_path](https://github.com/joblib/joblib/blob/master/joblib/_store_backends.py#L245). But, it requires one extra code line.

kirilldolmatov committed:
update request according to comments

1. Made location path OS independent 
2. Add `clear_location()` for full removing location folder

thomasjpfan commented:
This PR adds a `'cachedir'` into the examples directory by directly following the joblib doc and then removes `'cachedir'`. What was the issue with using a temporary folder for caching?

kirilldolmatov commented:
The issue will be the same: it's also possible to use a temporary folder (from `mkdtemp`):
```python
cachedir = mkdtemp()
memory = Memory(location=cachedir, verbose=10)
memory.clear(warn=False)
memory.store_backend.clear_location(cachedir)
```

thomasjpfan commented:
Does

```py
memory.clear(warn=False)
memory.store_backend.clear_location(cachedir)
```

do something that `rmtree(cachedir)` fails to do?

glemaitre commented:
@thomasjpfan Both solutions are equivalent. We either use the standard lib or joblib to achieve the same result. I am +0.5 for the change (it is nice that people find out about the joblib.Memory feature).

kirilldolmatov commented:
My initial PR replaced shutil and tempfile libs with joblib. Such solution looks simplifier, because we avoid import two extra libs and use only joblib. But, there is side effect - directory will remain after removing cache files. Therefore I add extra one code line, that removes this directory. Moreover, such solution allows use `mkdtemp`, if user doesn't like to create folder in directly way.

glemaitre commented on examples/compose/plot_compare_reduction.py:
```diff
@@ -119,7 +117,8 @@
 # Delete the temporary cache before exiting
-rmtree(cachedir)
+memory.clear(warn=False)
+memory.store_backend.clear_location(location)
```
IMO this is not a good idea. `store_backend` is not documented publicly. I would use the `rmtree` to remove the folder. In addition, I would add a comment that `memory.clear()` remove the cache used for the computation and `rmtree()` will remove the folder where the cache was localized.

kirilldolmatov replied:
I'd rather agree with you. I add `clear_location(location)` in my second commit for removing location directory after discussions above. IMHO, it isn't necessary to remove the folder, because it can be use in the future. If we're willing to accept this, than my first commit is actual. In another case, it might be better to use `rmtree `. My original wish was to replace three libs with one.

glemaitre replied:
> IMHO, it isn't necessary to remove the folder, because it can be use in the future

In general, I would agree. However, we don't want to have a remaining folder after the example execution. That's why, I would make a comment before the `rmtree` explaining why there is this call and that in practice, you probably don't need it.

> My original wish was to replace three libs with one.

It does not matter since they come in the standard library of Python

kirilldolmatov replied:
OK, I've made new commit.

kirilldolmatov committed:
use rmtree for removing folder

glemaitre reviewed (approved):
If you could just add a small comment, something like: `we remove the temporary directory earlier created (optional)`.

Otherwise LGTM

kirilldolmatov commented on examples/compose/plot_compare_reduction.py:
```diff
@@ -118,7 +119,7 @@
 # Delete the temporary cache before exiting
 memory.clear(warn=False)
-memory.store_backend.clear_location(location)
+rmtree(location)
```
remove the temporary directory earlier created (optional)

rth commented:
Merging, thanks!
